gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspConfigurations.kt

```diff
@@ -0,0 +1,136 @@
+package com.google.devtools.ksp.gradle
+
+import org.gradle.api.Project
+import org.gradle.api.artifacts.Configuration
+import org.jetbrains.kotlin.gradle.dsl.*
+import org.jetbrains.kotlin.gradle.plugin.*
+import org.jetbrains.kotlin.gradle.plugin.mpp.KotlinJvmAndroidCompilation
+
+/**
+ * Creates and retrieves ksp-related configurations.
+ */
+class KspConfigurations(private val project: Project) {
+    companion object {
+        private const val PREFIX = "ksp"
+    }
+
+    // "ksp" configuration, applied to the "main" source set of all targets.
+    private val rootMainConfiguration = project.configurations.create(PREFIX)
+
+    // Store all ksp configurations for quick retrieval.
+    private val kotlinConfigurations = mutableMapOf<KotlinSourceSet, Configuration>()
+    private val androidConfigurations = mutableMapOf<String, Configuration>()
+
+    @OptIn(ExperimentalStdlibApi::class)
+    private fun <T : Any> saveConfiguration(
+        owner: T,
+        parent: Configuration?,
+        name: String,
+        cache: MutableMap<T, Configuration>
+    ): Configuration {
+        val configName = PREFIX + name.replaceFirstChar { it.uppercase() }
+        val config = if (configName == parent?.name) {
+            // trying to add config with same parent name. Can happen, for example
+            // with ksp<Target> when <Target> is "". Just use parent.
+            parent
+        } else {
+            // maybeCreate to be future-proof, but we should never have a duplicate with current logic
+            project.configurations.maybeCreate(configName).apply {
+                if (parent != null) extendsFrom(parent)
```

Comment:
This is needed for ksp configurations that are not related to a source set. Currently there's one case where this happens - the root `ksp` configuration in a multiplatform project. But I think there might be more in the future, for example it wouldn't be bad to add `kspTest` in multiplatform and make it a parent of all test sets.

Reply:
All configurations should be associated to a source set, so that being said we probably do not want `ksp` in multiplatform project, it should always be `ksp<Target><SourceSetName>` or `ksp<Target>` (for the main source set of a target).

Reply:
I can remove `ksp` for multiplatform but it's a breaking change, a very breaking one I would say 😄 it'd break all MP tests for example. What do you think?

Reply:
It is definitely a breaking change, but the current situation creates more confusion than benefit.

Adding multiple artifacts to `ksp` configuration (across source sets) impacts all of them. In contrast, `implementation` under MPP `sourceSets.jvmMain.dependencies` actually maps to configuration `implementationJvm` so there are no conflicts.

Reply:
That was a feature in my mind. But I get your point, I will remove it. In the future I think we could add `KotlinDependenciesHandler.ksp()` extensions to be used in KMP just like implementation in your example - mapping to the appropriate `ksp<SourceSet>` configuration.